gamla/graph.py

```diff
@@ -8,30 +8,66 @@
 from gamla import functional, functional_generic
 
 
+@toolz.curry
+def is_better_node(node: Any, seen: Dict, key: Callable):
+    old_node = seen[key(node)]
+    return node > old_node
+
+
+@toolz.curry
+def save_to_dict(element: Any, dictionary: Dict, func: Callable):
+    dictionary[func(element)] = element
+
+
 @toolz.curry
 def graph_traverse(
-    source: Any, get_neighbors: Callable, key: Callable = toolz.identity
+    source: Any,
+    get_neighbors: Callable,
+    key: Callable = toolz.identity,
+    keep_best: bool = False,
+    seen_dict: Dict = {},
 ) -> Iterable:
-    yield from graph_traverse_many([source], get_neighbors=get_neighbors, key=key)
+    remember = save_to_dict(dictionary=seen_dict, func=key)
+    is_seen = toolz.compose_left(key, seen_dict.__contains__)
+    better_node = is_better_node(seen=seen_dict, key=key)
+
+    yield from graph_traverse_many(
+        [source],
+        get_neighbors=get_neighbors,
+        keep_best=keep_best,
+        is_seen=is_seen,
+        remember=remember,
+        better_node=better_node,
+    )
 
 
 @toolz.curry
 def graph_traverse_many(
-    sources: Any, get_neighbors: Callable, key: Callable = toolz.identity
+    sources: Any,
+    is_seen: Callable,
+    keep_best: bool,
+    remember: Callable,
+    get_neighbors: Callable,
+    better_node: Callable,
 ) -> Iterable:
     """BFS over a graph, yielding unique nodes.
 
     Note: `get_neighbors` must return elements without duplicates."""
     queue = [*sources]
-    seen = set(map(key, queue))
+    for element in queue:
+        remember(element)
 
     while queue:
         current = queue.pop()
         yield current
         for node in get_neighbors(current):
-            if key(node) not in seen:
-                seen.add(key(node))
+            if not is_seen(node):
+                remember(node)
                 queue = [node] + queue
+            elif keep_best:
```

Comment:
these are implementation details of is_seen / remember, implement them outside the bfs